File: Boyer-Moore/boyer-moore.cpp
#include <algorithm>
#include <iostream>
#include <map>
#include <string>
#include <utility>
#include <vector>

std::vector<int> Z_algo(std::string S){
    int n = S.size();
    int r = 0;
    int l = 0;
    std::vector<int> Z(n,0);
    Z[0] = n;
    for(int k = 1; k < n; k++){
        if (k > r) {
            int z_k = 0;
            while (k+z_k < n && S[z_k]==S[k+z_k]) z_k++;
            Z[k] = z_k;
            if (z_k > 0) {
                r = k + z_k - 1;
                l = k;
            }
        }
        else {
            int k2 = k-l;
            if (Z[k2] < r-k+1) Z[k] = Z[k2];
            else {
                int z_k = r-k+1;
                while (k+z_k < n && S[z_k]==S[k+z_k]) z_k++;
                Z[k] = z_k;
                r = k + z_k - 1;
                l = k;
            }
        }
    }
    return Z;
}

std::pair<std::vector<int>,std::vector<int>> precompute_L_and_l(std::string P){
    std::string PR = P;
    std::reverse(std::begin(PR), std::end(PR));
    std::vector<int> N = Z_algo(PR);
    std::reverse(std::begin(N), std::end(N));
    int n = N.size();
    
    std::vector<int> L(n,0);
    std::vector<int> l(n,0);
    for(int j = 0; j < n; j++) {
        int i = n - N[j];
        L[i] = j+1;
        if (N[j]==j+1) l[n-j-1] = j+1;
    }
    for(int i = n-2; i >= 0; i--) l[i] = std::max(l[i+1], l[i]);
    return std::make_pair(L,l);
}

std::map<char,std::vector<int>> extended_bad_rule(std::string P){
    std::map<char,std::vector<int>> R;
    int n = P.size();
    for(int i = 0; i < n; i++){
        if (R.find(P[i]) == R.end()){
            R[P[i]] =  {};
        }
        R[P[i]].push_back(i);
    }
    return R;
}

int good_rule(int i, std::vector<int> &L, std::vector<int> &l){
    int n = L.size();
    if (L[i] > 0) return n - L[i];
    else return n - l[i];
}

int bin_search(int a, int b, int i, std::vector<int> R){
    if (a == b) return R[a];
    int m = (a+b)/2;
    if (R[m+1] < i) return bin_search(m+1, b, i, R);
    else return bin_search(a, m, i, R);
}

int bad_rule(int i, char P_h, std::map<char,std::vector<int>> &R, int n){
    if  (R.find(P_h) == R.end()) return n-i+1;
    if (R[P_h][0] > i) return n-i+1;
    return i - bin_search(0,R[P_h].size()-1,i, R[P_h]) ;
}

std::vector<int> boyer_moore(std::string P, std::string T){
    std::vector<int> res;
    std::vector<int> L,l;
    std::tie(L,l) = precompute_L_and_l(P);
    std::map<char,std::vector<int>> R = extended_bad_rule(P);

    int n = P.size();
    int m = T.size();
    int k = n-1;
    while (k < m) {
        int i = n-1;
        int h = k;
        while ((i >= 0) && (P[i] == T[h])) {
            i--; h--;
        }
        if (i == -1){
            res.push_back(k);
            k = k + n - l[1];
        }
        else{
            k = k + std::max(good_rule(i,L,l),bad_rule(i,T[h],R,n));
        }
    }
    return res;
}

int main(){
    std::string P,T;
    getline(std::cin,P);
    getline(std::cin,T);
    std::vector<int> res = boyer_moore(P,T);
    for(int i = 0; i < res.size(); i++) std::cout << res[i] << " ";
    std::cout << std::endl;
    return 0;
}

File: Boyer-Moore/boyer_moore.py
def preprocessing_extended_bad_rule(P):
    n = len(P)
    R = {}
    for i in range(n):
        if P[i] not in R:
            R[P[i]] = []
        R[P[i]].append(i)
    return R

def N_preprocessing(P):
    P = P[::-1]
    n = len(P)
    N = [0 for i in range(n)]
    r = l = 0
    for k in range(1,n):
        if r < k :
            n_k = 0
            while k + n_k < n and P[n_k]==P[n_k+k]:
                n_k +=1
            N[k] = n_k
            if n_k > 0:
                l = k
                r = k + n_k - 1
        else:
            k2 = k-l
            if N[k2] < r - k + 1:
                N[k] = N[k2]
            else:
                n_k = r-k+1
                while k + n_k < n and P[n_k]==P[n_k+k]:
                    n_k +=1
                N[k] = n_k
                l = k
                r = k + n_k - 1
    N.reverse()
    return N

def L_and_l_preprocessing(P):
    n = len(P)
    N = N_preprocessing(P)
    L = [0 for i in range(n)]
    l = [0 for i in range(n)]
    for j in range(n-1):
        i = n - N[j]
        if i < n :
            L[i] = j+1
        if N[j] == j+1 :
            l[n-j-1] = j+1
    for i in range(n-2,-1,-1):
        l[i] = max(l[i],l[i+1])
    l[n-1] = n-1
    return L,l

def boyer_moore(P,T):
    res = []
    n = len(P)
    m = len(T)
    R = preprocessing_extended_bad_rule(P)
    L,l = L_and_l_preprocessing(P)

    k = n-1
    while k < m:
        h = k
        i = n-1
        while i >= 0 and T[h]==P[i]:
            i -= 1
            h -= 1
        if i == -1 :
            res.append(k-n+1)
            if len(l) >1 :
                k = k + n - l[1]
            else :
                k = k + 1
        else :
            good_rule = 1
            if i+1 == n:
                pass
            elif L[i+1] > 0 :
                good_rule = n - L[i+1]
            else :
                good_rule = n - l[i+1]

            bad_rule = 1
            if T[h] not in R:
                bad_rule = i+1
            elif R[T[h]][0] > i :
                bad_rule = i+1
            else :
                a = 0
                b = len(R[T[h]])-1
                while a != b :
                    mil = (a+b)//2
                    if R[T[h]][mil+1] < i :
                        a = mil+1
                    else :
                        b = mil
                bad_rule = i-R[T[h]][a]

            k = k + max(bad_rule,good_rule)

    return res
